packages/lodestar/src/network/gossip/gossipsub.ts

```diff
@@ -0,0 +1,118 @@
+import Gossipsub, {IGossipMessage, Registrar, Options} from "libp2p-gossipsub";
+import {IGossipMessageValidator, GossipObject, GossipMessageValidatorFn} from "./interface";
+import {getGossipTopic, deserializeGossipMessage} from "./utils";
+import {GossipEvent} from "./constants";
+import {Attestation, AggregateAndProof, AttesterSlashing, ProposerSlashing, 
+  SignedVoluntaryExit, SignedBeaconBlock} from "@chainsafe/eth2.0-types";
+import {IBeaconConfig} from "@chainsafe/eth2.0-config";
+import {utils} from "libp2p-pubsub";
+import {ATTESTATION_SUBNET_COUNT} from "../../constants";
+
+/**
+ * This validates messages in Gossipsub and emit the transformed messages.
+ * We don't want to double deserialize messages for performance benefit.
+ */
+export class LodestarGossipsub extends Gossipsub {
+  private transformedObjects: Map<string, GossipObject>;
+  private config: IBeaconConfig;
+  private validator: IGossipMessageValidator;
+
+  constructor (config: IBeaconConfig, validator: IGossipMessageValidator, peerInfo: PeerInfo,
+    registrar: Registrar, options: Options = {}) {
+    super(peerInfo, registrar, options);
+    this.transformedObjects = new Map();
+    this.config = config;
+    this.validator = validator;
+  }
+
+  public async validate(rawMessage: IGossipMessage): Promise<boolean> {
+    const message: IGossipMessage = utils.normalizeInRpcMessage(rawMessage);
+    if (!message.topicIDs) {
+      return false;
+    }
+
+    let isValid = true;
+    let transformedObj: GossipObject;
+    for (const topic of message.topicIDs) {
```

Comment:
Should we discard messages with more than one topic id?

Reply:
yeah probably because when we only publish single topic id but let's double check.
@wemeetagain wdyt?

Reply:
asked this question on eth2 discord, here's proto's answer:
Either gracefully process for each topic, or discard (with an error log) to catch accidental multiple topic IDs

Reply:
I'll go with the 2nd option to make it simple.